libcudacxx/include/cuda/std/__cccl/dialect.h

```diff
@@ -126,4 +126,44 @@
 #  define _CCCL_GLOBAL_CONSTANT _CCCL_INLINE_VAR constexpr
 #endif // __CUDA_ARCH__
 
+////////////////////////////////////////////////////////////////////////////////
+// _CCCL_TEMPLATE
+// Usage:
+//   _CCCL_TEMPLATE(typename A, typename _Bp)
+//     _CCCL_REQUIRES( Concept1<A> _CCCL_AND Concept2<_Bp>)
+//   void foo(A a, _Bp b)
+//   {}
+
+// Barebones enable if implementation to use outside of cuda::std
+template <bool>
+struct __cccl_select
+{};
+
+template <>
+struct __cccl_select<true>
+{
+  template <class _Tp>
+  using type = _Tp;
+};
+
+template <bool _Bp, class _Tp = void>
+using __cccl_enable_if_t = typename __cccl_select<_Bp>::template type<_Tp>;
+
+template <class _Tp, bool _Bp>
+using __cccl_requires_t = typename __cccl_select<_Bp>::template type<_Tp>;
+
+#if (defined(__cpp_concepts) && _CCCL_STD_VER >= 2020)
+#  define _CCCL_TEMPLATE(...)               template <__VA_ARGS__>
+#  define _CCCL_REQUIRES(...)               requires __VA_ARGS__
+#  define _CCCL_AND                         &&
+#  define _CCCL_TRAILING_REQUIRES_AUX_(...) requires __VA_ARGS__
+#  define _CCCL_TRAILING_REQUIRES(...)      ->__VA_ARGS__ _CCCL_TRAILING_REQUIRES_AUX_
+#else // ^^^ __cpp_concepts ^^^ / vvv !__cpp_concepts vvv
+#  define _CCCL_TEMPLATE(...)               template <__VA_ARGS__
+#  define _CCCL_REQUIRES(...)               , bool _CCCL_true_ = true, __cccl_enable_if_t < __VA_ARGS__ && _CCCL_true_, int > = 0 >
+#  define _CCCL_AND                         &&_CCCL_true_, int > = 0, __cccl_enable_if_t <
+#  define _CCCL_TRAILING_REQUIRES_AUX_(...) , __VA_ARGS__ >
+#  define _CCCL_TRAILING_REQUIRES(...)      ->__cccl_requires_t < __VA_ARGS__ _CCCL_TRAILING_REQUIRES_AUX_
+#endif // !__cpp_concepts
+
```

Comment:
seems to me that `__concept_macros.h` was the right place for these. i would leave them there, just rename them.

Reply:
For me there is a set of globally available features that I want to use in all of CCCL

The concept macros are one of them. I would not want to port `_LIBCUDACXX_CONCEPT` and friends but this is something that is universally useful and should be commonly available

Reply:
Moved back to concept_macros